compiler/vm/cbackend2.nim

```diff
@@ -0,0 +1,1097 @@
+      if it.kind in routineKinds:
+        nextProcs.add it
+
+    drain(c, g.config, env, procImpls, nextProcs, nextProcs2, seenProcs)
+
+  # generate all deferred symbols and declarations. ``c.defSyms`` should not
+  # be used past this point
+  flush(c.defSyms, g.cache, env.syms)
+
+  # XXX: it would be better to also set the type during
+  #      ``DeferredSymbols.flush``. But then the procedure needs access to a
+  #      ``DeferredTypeGen`` object - which is a bit awkward, since
+  #      ``DeferredTypeGen.flush`` requires a ``DeferredSymbol`` object
+  for id, s in env.syms.msymbols:
+    if (let orig = env.syms.orig.getOrDefault(id); orig != nil):
+      s.typ = c.types.requestType(orig.typ)
+
+  # XXX: this is a problem. Resolving the type-bounds operation requires
+  #      all alive types being present, but that isn't the case yet since
+  #      the final flush hasn't happenend, but for that we need to call
+  #      ``finish``, but we can't since we're still adding procedures
+  resolveTypeBoundOps(passEnv, g, c.types, c.procs)
+
+  block:
+    # translate the literal data from it's ``PNode``-based representation
+    for id, data in c.constData.pairs:
+      assert env.syms[id].kind == skConst
+      env.syms.setData(id): add(c.data, c.procs, g.config, data)
+
+    reset c.constData # no longer needed
+
+  c.procs.finish(c.types, g.cache)
+
+  block:
+    # flush all remaining deferred types. ``c.symEnv`` was consumed when
+    # flushing it, so we create a temporary new ``DeferredSymbol`` object here
+    var defSyms = initDeferredSyms(env.syms)
+    c.types.flush(env.types, defSyms, g.config)
+    c.types.flush2(env.types, c.data, g.config)
+
+    flush(defSyms, g.cache, env.syms)
+
+  # replace all placeholder type IDs
+  finishTypes(passEnv, c.types, procImpls, c.procs, env.syms)
+
+  processObjects(passEnv, g.cache, env.types, env.syms, c.types.objects)
+
+  swap(c.procs, env.procs)
+  swap(c.data, env.data)
+
+  let objects = move c.types.objects
+
+  # TODO: ``generateCode`` needs to be split up. Freeing the ``TCtx`` should
+  #       happen automatically as part of stack-frame clean up
+  reset(c) # we no longer need the irgen context
+
+  block:
+    # all alive globals are collected by now - register them with their
+    # owning module
+
+    for id in env.syms.items:
+      let s = env.syms[id]
+      case s.kind
+      of skVar, skLet, skForVar:
+        let mIdx = mlist[].lookupModule(env.syms.orig[id])
+        modules[mIdx].syms.add(id)
+
+      else:
+        discard
+
+  # HACK: in the context of RTTI, closure types are represented via a "fake"
+  #       tuple type (one not matching the type they're lowered to).
+  # the later RTTI setup logic generation requires the types it processes to
+  # have corresponding entries in the `tfInfo` and `gcInfo` lookup tables,
+  # so the type has to be added to the environment before computing the
+  # tables
+  let fakeClosure = genFakeClosureType(env.types, passEnv)
```

Comment:
Could you elaborate on that? I'm not sure I understand.

I am/was under the impression that sem should be unaware of a targets implementation details. The `fakeClosure` type is a slight hack that is only needed for the `TNimType`-based RTTI, which is only used for the C and C++ targets if `optTinyRtti` is not enabled.

Reply:
Yup, I think sem should never ever ask which backend and then make a decision, so your absolutely correct about that.

What I meant was when the compiler is choosing a backend and then sets up the passes, it will configure sem to apply some optional transforms. These optional transforms are uniform in implementation but aren't universally enabled for all backends (at time of pass setup).